### PR TITLE
Change humanize() granularity="auto" limits

### LINKED ISSUE #848: humanize() rounding problem
## Issue Description
Not sure if this is wanted or a bug. `humanize()` returns "a week ago" for deltas < 7 days.
For example:
```
>>> arrow.utcnow().shift(days=-6, hours=-9).humanize()
'6 days ago'
>>> arrow.utcnow().shift(days=-6, hours=-10).humanize()
'a week ago'
```
I would expect "a week ago" to be returned after 7 days have passed.

## System Info

- 🖥 **OS name and version**: Fedora 32
- 🐍 **Python version**: Python 3.8.5
- 🏹 **Arrow version**: arrow 0.16.0

### PR DESCRIPTION
**Pull request by systemcatch**
## Pull Request Checklist

Thank you for taking the time to improve Arrow! Before submitting your pull request, please check all *appropriate* boxes:

- [ ] 🧪  Added **tests** for changed code.
- [x] 🛠️  All tests **pass** when run locally (run `tox` or `make test` to find out!).
- [x] 🧹  All linting checks **pass** when run locally (run `tox -e lint` or `make lint` to find out!).
- [ ] 📚  Updated **documentation** for changed code.
- [ ] ⏩  Code is **up-to-date** with the `master` branch.

If you have *any* questions about your code changes or any of the points above, please submit your questions along with the pull request and we will try our best to help!

## Description of Changes

First attempt at updating the `humanize()` auto limits to avoid weird results.

Need to check failing tests and finish the month/months limit.

closes #848

### CODE REVIEW

**Review by krisfremen — APPROVED** [review]
Only one comment, otherwise looks good. :+1:

File: arrow/arrow.py [hunk 1]
```diff
@@ -1012,42 +1012,42 @@ def humanize(
...
                     seconds = sign * delta
                     return locale.describe(
                         "seconds", seconds, only_distance=only_distance
                     )
 
-                elif diff < 90:
+                elif diff < 120:
                     return locale.describe("minute", sign, only_distance=only_distance)
-                elif diff < 2700:
+                elif diff < 3600:
```
[the remark this answers is not in the document]
Comment (systemcatch) [reply arrow/arrow.py:1023]: @krisfremen I've been looking over the code and saw these class attributes (https://github.com/arrow-py/arrow/blob/master/arrow/arrow.py#L70), we use them when a granularity is specified, could we just use them again here?

[the remark this answers is not in the document]
Comment (jadchaar) [reply arrow/arrow.py:1023]: I think using static constants might be ideal here.

File: arrow/util.py [hunk 2]
```diff
@@ -9,6 +10,18 @@
...
 
 
+def seconds_in_month(dt):
+
+    year = dt.year
+    month = dt.month
+
+    days_in_month = calendar.monthrange(year, month)[1]
+
+    seconds = days_in_month * 86400
```
Comment (jadchaar) [inline arrow/util.py:20]: We should probably get 86400 into a variable rather than having a magic constant.

**Review by jadchaar — COMMENTED** [review]
Looks good but I think we should get rid of the magic constants and instead refer to the static class constants.

Also are you planning to refactor and break up the humanize function in another PR?

File: arrow/arrow.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (systemcatch) [reply arrow/arrow.py:1023]: `_SECS_PER_YEAR` causes some issues but the others are fine.

File: arrow/util.py (hunk 2 above)
Comment (systemcatch) [reply arrow/util.py:20]: Yeah to be honest I think this function has some problems. Because we are calculating total seconds for the current month that could mean we under/over estimate the delta for the follwing month

**Comment by systemcatch** [comment]
> Looks good but I think we should get rid of the magic constants and instead refer to the static class constants.

👍 

> Also are you planning to refactor and break up the humanize function in another PR?

Yeah that can wait.

File: arrow/util.py (hunk 2 above)
Comment (systemcatch) [reply arrow/util.py:20]: Yeah I'm thinking there's a few issues with this method.

**Comment by systemcatch** [comment]
Right so I've made changes based on both your suggestions, however there are still some problems to work out.

For example:
```shell
(arrow) chris@ThinkPad:~/arrow$ python
Python 3.8.3 (default, Jul  7 2020, 18:57:36) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import arrow
>>> dt=arrow.utcnow()
>>> later=dt.shift(years=1)
>>> later.humanize(dt)
'in a year'
>>> later.humanize(dt, granularity="year")
'in 0 years'
```
This happens because `_SECONDS_PER_YEAR` is `60*60*24*365.25` whereas I've used `60*60*24*365`. I tend to think it's not worth worrying about leap years in humanize. There will be a similar problem with months which I don't think is such an issue.

**Comment by jadchaar** [comment]
> 
> 
> Right so I've made changes based on both your suggestions, however there are still some problems to work out.
> 
> For example:
> 
> ```shell
> (arrow) chris@ThinkPad:~/arrow$ python
> Python 3.8.3 (default, Jul  7 2020, 18:57:36) 
> [GCC 9.3.0] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import arrow
> >>> dt=arrow.utcnow()
> >>> later=dt.shift(years=1)
> >>> later.humanize(dt)
> 'in a year'
> >>> later.humanize(dt, granularity="year")
> 'in 0 years'
> ```
> 
> This happens because `_SECONDS_PER_YEAR` is `60*60*24*365.25` whereas I've used `60*60*24*365`. I tend to think it's not worth worrying about leap years in humanize. There will be a similar problem with months which I don't think is such an issue.

Do you think we should define a class in constants called Humanize constants? Or even a static inner class of the Arrow object? Because I don't think these .25 adjustments for leap years should be used for humanize. Thoughts?

**Comment by systemcatch** [comment]
> Do you think we should define a class in constants called Humanize constants? Or even a static inner class of the Arrow object? Because I don't think these .25 adjustments for leap years should be used for humanize. Thoughts?

Maybe we just change `_SECS_PER_YEAR` to `60*60*24*365`? It's not used anywhere else apart from humanize.

**Comment by jadchaar** [comment]
> > Do you think we should define a class in constants called Humanize constants? Or even a static inner class of the Arrow object? Because I don't think these .25 adjustments for leap years should be used for humanize. Thoughts?
> 
> Maybe we just change `_SECS_PER_YEAR` to `60*60*24*365`? It's not used anywhere else apart from humanize.

Yeah I think it makes most sense to accommodate it for where it is being used. What do you think?

**Comment by systemcatch** [comment]
@jadchaar I've changed the value to `60*60*24*365`

File: arrow/constants.py [hunk 3]
```diff
@@ -7,3 +7,19 @@
 MAX_TIMESTAMP = 253402318799.0
 MAX_TIMESTAMP_MS = MAX_TIMESTAMP * 1000
 MAX_TIMESTAMP_US = MAX_TIMESTAMP * 1000000
+
+# only some of these are used, the others are already covered by Arrow class attributes.
+humanize_auto_limits = {
```
Comment (jadchaar) [inline arrow/constants.py:12]: We should probably capitalize this since its a constant.

File: tests/test_arrow.py [hunk 4]
```diff
@@ -1954,7 +1954,7 @@ def test_years(self):
 
         result = arw.humanize(self.datetime)
 
-        assert result == "in 2 years"
+        assert result == "in a year"
```
Comment (jadchaar) [inline tests/test_arrow.py:1958]: Should we adjust tests to account for the newly updated bounds? E.g. account for the edge cases?

File: arrow/constants.py [hunk 5]
```diff
@@ -7,3 +7,19 @@
...
+    "minute": 120,
+    "minutes": 3600,
+    "hour": 7200,
+    "hours": 86400,
+    "day": 172800,
+    "days": 604800,
+    "week": 1209600,
+    "month": 3888000,
+    "months": 31536000,
+    "year": 63072000,
```
Comment (jadchaar) [inline arrow/constants.py:24]: Should we show the multiplications rather than using magic variables? Or do you think just hardcoding the constants is enough?

**Review by jadchaar — COMMENTED** [review]
A few comments, but looking good!

File: arrow/constants.py (hunk 5 above)
Comment (krisfremen) [reply arrow/constants.py:24]: I'd concur with the multiplications. It would be easier to visualise.

Comment (systemcatch) [reply arrow/constants.py:24]: Good idea I'll change it.

File: tests/test_arrow.py (hunk 4 above)
Comment (systemcatch) [reply tests/test_arrow.py:1958]: I can add some extra tests for edge cases.

File: arrow/arrow.py [hunk 6]
```diff
@@ -1012,42 +1012,42 @@ def humanize(
...
-                elif diff < 5400:
+                elif diff < self._SECS_PER_HOUR * 2:
                     return locale.describe("hour", sign, only_distance=only_distance)
-                elif diff < 79200:
-                    hours = sign * int(max(delta / 3600, 2))
+                elif diff < self._SECS_PER_DAY:
+                    hours = sign * int(max(delta / self._SECS_PER_HOUR, 2))
                     return locale.describe("hours", hours, only_distance=only_distance)
 
                 # anything less than 48 hours should be 1 day
```
Comment (jadchaar) [inline arrow/arrow.py:1035]: Should we put more of these comments in each case? E.g. anything less than 2 weeks is a week? Or do you think the code is clear enough and we can just remove this comment?

File: arrow/arrow.py [hunk 7]
```diff
@@ -72,7 +72,7 @@ class Arrow(object):
     _SECS_PER_DAY = float(60 * 60 * 24)
     _SECS_PER_WEEK = float(60 * 60 * 24 * 7)
     _SECS_PER_MONTH = float(60 * 60 * 24 * 30.5)
-    _SECS_PER_YEAR = float(60 * 60 * 24 * 365.25)
+    _SECS_PER_YEAR = float(60 * 60 * 24 * 365)
```
Comment (jadchaar) [inline arrow/arrow.py:75]: Should we put a comment similar to the weeks so that we remember to revisit year humanization during leap years?

File: tests/test_arrow.py [hunk 8]
```diff
@@ -1776,7 +1776,7 @@ def test_multiple_granularity(self):
             self.now.humanize(
                 later108onlydistance, only_distance=True, granularity=["year", "second"]
             )
-            == "3 years and 5327200 seconds"
+            == "3 years and 5392000 seconds"
```
Comment (jadchaar) [inline tests/test_arrow.py:1779]: Why did this change? Is it because of the 365.25 turning to 365?

**Review by jadchaar — APPROVED** [review]
A few comments, but overall it looks good to merge!

File: arrow/arrow.py (hunk 6 above)
Comment (systemcatch) [reply arrow/arrow.py:1035]: I think we can remove it.

File: tests/test_arrow.py (hunk 8 above)
Comment (systemcatch) [reply tests/test_arrow.py:1779]: Yeah man, 3 years of 0.25 is 64800 seconds. I'll add a comment.

**Comment by systemcatch** [comment]
@jadchaar I made your suggested changes, we should be good to go here.
